drivers/eeprom/eeprom_mchp_xec.c

```diff
@@ -0,0 +1,326 @@
+	}
+	regs->execute = temp;
+}
+
+static void eeprom_xec_data_buffer_read(struct eeprom_xec_regs * const regs,
+					uint8_t transfer_size, uint8_t *destination_ptr)
+{
+	uint8_t count;
+
+	if (transfer_size <= XEC_EEPROM_PAGE_SIZE) {
+		for (count = 0; count < transfer_size; count++) {
+			*destination_ptr = regs->buffer[count];
+			destination_ptr++;
+		}
+	}
+}
+
+static void eeprom_xec_data_buffer_write(struct eeprom_xec_regs * const regs,
+							uint8_t transfer_size, uint8_t *source_ptr)
+{
+	uint8_t count;
+
+	if (transfer_size <= XEC_EEPROM_PAGE_SIZE) {
+		for (count = 0; count < transfer_size; count++) {
+			regs->buffer[count] = *source_ptr;
+			source_ptr++;
+		}
+	}
+}
+
+static void eeprom_xec_wait_transfer_compl(struct eeprom_xec_regs * const regs)
+{
+	uint8_t sts = 0;
+	uint8_t retry_count = 0;
+
+	k_sleep(K_USEC(XEC_EEPROM_DELAY_US));
+
+	do {
+		if (retry_count >= XEC_EEPROM_XFER_COMPL_RETRY_COUNT) {
+			LOG_ERR("XEC EEPROM retry count exceeded");
+			break;
+		}
+		k_sleep(K_USEC(XEC_EEPROM_DELAY_BUSY_POLL_US));
+
+		sts = XEC_EEPROM_STS_TRANSFER_COMPL & regs->status;
+		retry_count++;
+
+	} while (!sts);
+
+	/* Clear the appropriate status bits */
+	regs->status = XEC_EEPROM_STS_TRANSFER_COMPL;
+}
+
+static void eeprom_xec_wait_write_compl(struct eeprom_xec_regs * const regs)
+{
+	uint8_t sts = 0;
+	uint8_t retry_count = 0;
+
+	do {
+		if (retry_count >= XEC_EEPROM_XFER_COMPL_RETRY_COUNT) {
+			LOG_ERR("XEC EEPROM retry count exceeded");
+			break;
+		}
+
+		regs->buffer[0] = 0;
+
+		/* Issue the READ_STS command */
+		regs->execute = XEC_EEPROM_EXC_CMD_READ_STS;
+
+		eeprom_xec_wait_transfer_compl(regs);
+
+		sts = regs->buffer[0] & (XEC_EEPROM_STS_BYTE_WIP |
+							XEC_EEPROM_STS_BYTE_WENB);
+
+		retry_count++;
+
+	} while (sts);
+}
+
+static void eeprom_xec_data_read_32_bytes(struct eeprom_xec_regs * const regs,
+							uint8_t *buf, size_t len, off_t offset)
+{
+	/* Issue the READ command to transfer buffer to EEPROM memory */
+	eeprom_xec_execute_reg_set(regs, len, XEC_EEPROM_EXC_CMD_READ, offset);
+
+	/* Wait until the read operation has completed */
+	eeprom_xec_wait_transfer_compl(regs);
+
+	/* Read the data in to the software buffer */
+	eeprom_xec_data_buffer_read(regs, len, buf);
+}
+
+static void eeprom_xec_data_write_32_bytes(struct eeprom_xec_regs * const regs,
+							uint8_t *buf, size_t len, off_t offset)
+{
+	uint16_t sz, rem_bytes;
+
+	sz = offset % XEC_EEPROM_PAGE_SIZE;
+
+	/* If EEPROM Addr is not on page boundary */
+	if (sz != 0) {
+		/* Check if we are crossing page boundary */
+		if ((sz + len) > XEC_EEPROM_PAGE_SIZE) {
+			rem_bytes = (XEC_EEPROM_PAGE_SIZE - sz);
+			/* Update the EEPROM buffer */
+			eeprom_xec_data_buffer_write(regs, rem_bytes, buf);
+
+			/* Issue the WRITE command to transfer buffer to EEPROM memory */
+			eeprom_xec_execute_reg_set(regs, rem_bytes,
+						XEC_EEPROM_EXC_CMD_WRITE, offset);
+
+			eeprom_xec_wait_transfer_compl(regs);
+
+			eeprom_xec_wait_write_compl(regs);
```

Comment:
`eeprom_xec_wait_write_compl` function already checks for `eeprom_xec_wait_transfer_compl`. So, do you think it's needed here too?

Reply:
yes it is required; the first transfer complete is checking if the WRITE command is complete; the second transfer complete (inside eeprom_xec_wait_write_compl) is checking if the READ_STS command complete.